pygfx/objects/_base.py

```diff
@@ -1,7 +1,31 @@
+import weakref
+
 from ..linalg import Vector3, Matrix4, Quaternion
 
 
-class WorldObject:
+class TrackableObject:
+    def __init__(self):
+        self._set_attr_callbacks = {}
+
+    def __setattr__(self, name, value):
+        super().__setattr__(name, value)
+        try:
+            callbacks = self._set_attr_callbacks[name]
+        except KeyError:
+            pass
+        else:
+            if not callbacks:
+                self._set_attr_callbacks.pop(name)  # Make it quicker next time
+            for cb in callbacks:
+                cb(self, name)
+
+    def _listen(self, name, fn):
+        # Private, because only intended for private use
+        callbacks = self._set_attr_callbacks.setdefault(name, weakref.WeakSet())
+        callbacks.add(fn)
```

Comment:
@Korijn What do you think of this? It's a base class for WorldObject, Material and Geometry. It allows the render functions to be notified when a specific attribute is set.

In practice, render functions currently listen for e.g. `wobject.geometry`, `wobject.material`, `geometry.posisions` etc, `material.map`.

Any other ideas/approaches?

Reply:
At first glance this looks good. I dislike that it's implemented through a base class (TrackableObject) but I understand how it's convenient wrt cleanup/garbage collection since the listener lifetimes are tied to the lifetime of the object. :) I made the same design decision in observ for the same reasons.

Give me some time to read up on solutions in other libraries so I can compare them to your approach.

Reply:
I think the approach here suffices. We'll see if we run into any limits in the future.

Reply:
Looking at what the renderers subscribe to, it seems to be safe to say that (apart from the geometry or material being reset) any attribute being set that is a resource (buffer of texture), requires a pipeline reset. Implementing this is easier and less intrusive - no callback stuff.

Reply:
The only case I can think of is e.g. a material with a property that causes the thing to be rendered in a different style, e.g. with a different shader. But one could argue that you'd need to implement a different Material subclass instead of using a property.